tools/cmd/scadctl/cmd/gardener.go

```diff
@@ -16,32 +43,512 @@ var scenarioDir string
+	default:
+		return nil, fmt.Errorf("unsupported gardener plane: %d", plane)
+	}
+
+	cmdStr := fmt.Sprintf("eval $(gardenctl kubectl-env bash) && %s > /dev/null && gardenctl kubectl-env bash", targetStr)
+	cmd := exec.CommandContext(ctx, "bash", "-c", cmdStr)
+	cmd.Env = append(os.Environ(), "GCTL_SESSION_ID=dev")
+
+	capturedOut, err := invokeCommand(cmd)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute gardenctl command: %w", err)
+	}
+
+	kubeConfigPath, err := extractKubeConfigPath(capturedOut)
+	if err != nil {
+		return nil, err
+	}
+
+	restCfg, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create rest.Config from kubeconfig %q: %w", kubeConfigPath, err)
+	}
+
+	return client.New(restCfg, client.Options{Scheme: scheme})
+}
+
+// ---------------------------------------------------------------------------------
+// Cluster Snapshot
+// ---------------------------------------------------------------------------------
+func (a *access) ListNodes(ctx context.Context, criteria mkapi.MatchCriteria) (nodes []corev1.Node, err error) {
+	var nodeList corev1.NodeList
+	err = a.shootClient.List(ctx, &nodeList, &client.ListOptions{
+		LabelSelector: criteria.LabelSelector,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if criteria.Names.Len() <= 0 {
+		return nodeList.Items, nil
+	}
+	for _, n := range nodeList.Items {
+		if criteria.Names.Has(n.Name) {
+			nodes = append(nodes, n)
+		}
+	}
+	return nodes, err
+}
+
+func (a *access) ListPods(ctx context.Context, criteria mkapi.MatchCriteria) (pods []corev1.Pod, err error) {
+	var podList corev1.PodList
+	err = a.shootClient.List(ctx, &podList, &client.ListOptions{
+		Namespace:     criteria.Namespace,
+		LabelSelector: criteria.LabelSelector,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if criteria.Names.Len() <= 0 {
+		return podList.Items, nil
+	}
+	for _, p := range podList.Items {
+		// TODO Check if needed: filter pods having scheduling gates
+		if criteria.Names.Has(p.Name) && len(p.Spec.SchedulingGates) != 0 {
+			pods = append(pods, p)
+		}
+	}
+	return pods, err
+}
+
+func (a *access) ListPriorityClasses(ctx context.Context) ([]schedulingv1.PriorityClass, error) {
+	var priorityClassList schedulingv1.PriorityClassList
+	err := a.shootClient.List(ctx, &priorityClassList)
+	return priorityClassList.Items, err
+}
+
+func (a *access) ListRuntimeClasses(ctx context.Context) ([]nodev1.RuntimeClass, error) {
+	var runtimeClassList nodev1.RuntimeClassList
+	err := a.shootClient.List(ctx, &runtimeClassList)
+	return runtimeClassList.Items, err
+}
+
+func (a *access) MakeCSIDriverVolumeMap(ctx context.Context) (map[string]int32, error) {
+	var csiNodeList storagev1.CSINodeList
+	err := a.shootClient.List(ctx, &csiNodeList)
+	if len(csiNodeList.Items) == 0 {
+		return nil, err
+	}
+
+	volMap := make(map[string]int32, 0)
+	for _, csiNode := range csiNodeList.Items {
+		for _, d := range csiNode.Spec.Drivers {
```

Comment:
Every `CSINodeDriver` also has `TopologyKeys` which as per the docs is used by the kube-scheduler. Can you check if these are indeed being used and if then that information should also be captured.

Reply:
Ok after some investigation it seems that the `kube-scheduler` will not use the TopologyKeys that are defined as part of `CSINode` object.  You can refer to https://kubernetes-csi.github.io/docs/topology.html to understand the significance of topology keys w.r.t CSIDriver.

`kube-scheduler` uses the following information to ensure that the pods having volume dependency are correctly scheduled:
* Uses `StorageClass` volumeBindingMode and allowedTopologies.
* Uses standard topology labels on Node objects.